### PR TITLE
test: add consumer test ( E2E )

### PR DESCRIPTION
**Pull request by idbeta**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [ ] Bugfix
- [ ] New feature provided
- [ ] Improve performance
- [x] E2E test

- Related issues

### CODE REVIEW

**Review by membphis — APPROVED** [review]

**Review by membphis — CHANGES_REQUESTED** [review]

File: api/test/e2e/consumer_test.go [hunk 1]
```diff
@@ -0,0 +1,252 @@
...
+
+	for _, tc := range tests {
+		testCaseCheck(tc)
+	}
+}
+
+func TestConsumer_delete_notexit_consumer(t *testing.T) {
+	tests := []HttpTestCase{
+		{
+			caseDesc:     "delete notexit consumer",
```
Comment (membphis) [inline api/test/e2e/consumer_test.go]: Please check your English

File: api/test/e2e/consumer_test.go [hunk 2]
```diff
@@ -0,0 +1,252 @@
...
+				 "plugins": {
+					 "key-auth": {
+						 "key": "auth-new"
+					 }
+				 },
+				 "desc": "test description"
+			 }`,
+			Headers:      map[string]string{"Authorization": token},
+			ExpectStatus: http.StatusBadRequest,
+			ExpectBody:   "scheme validate fail",
```
Comment (membphis) [inline api/test/e2e/consumer_test.go]: Need a check at the date plan.

File: api/test/e2e/consumer_test.go [hunk 3]
```diff
@@ -0,0 +1,214 @@
...
+				"username": "jack",
+				"labels": {
+					"build":"16",
+					"env":"production",
+					"version":"v2"
+				},
+				"plugins": {
+					"key-auth": {
+						"key": "auth-two"
+					}
```
Comment (membphis) [inline api/test/e2e/consumer_test.go:83]: This case is wrong.

Comment (membphis) [reply api/test/e2e/consumer_test.go:83]: You are using a invalid plugin name.

File: api/test/e2e/consumer_test.go [hunk 4]
```diff
@@ -0,0 +1,214 @@
...
+	time.Sleep(time.Duration(1000) * time.Millisecond)
+
+	//update the consumer with new desc
+	data = `{
+		"username":"jack",
+		"desc": "updated consumer"
+    }`
+	request, _ = http.NewRequest("PUT", basepath, strings.NewReader(data))
+	request.Header.Add("Authorization", token)
+	resp, _ = http.DefaultClient.Do(request)
```
Comment (membphis) [inline api/test/e2e/consumer_test.go:189]: After we deleted them, we need to make a confirm at the data plane.

**Review by membphis — CHANGES_REQUESTED** [review]

File: api/test/e2e/consumer_test.go [hunk 5]
```diff
@@ -0,0 +1,214 @@
...
+	
+	//get the consumer, save createtime and updatetime
+	request, _ = http.NewRequest("GET", basepath+"/jack", nil)
+	request.Header.Add("Authorization", token)
+	resp, _ = http.DefaultClient.Do(request)
+	respBody, _ = ioutil.ReadAll(resp.Body)
+	createtime := gjson.Get(string(respBody), "data.create_time")
+	updatetime := gjson.Get(string(respBody), "data.update_time")
+
+	//Wait one second for update time to be different
```
Comment (LiteSun) [inline api/test/e2e/consumer_test.go:179]: should keep the same comment style.

File: api/test/e2e/consumer_test.go [hunk 6]
```diff
@@ -0,0 +1,220 @@
...
+			ExpectStatus: http.StatusOK,
+		},
+	}
+
+	for _, tc := range tests {
+		testCaseCheck(tc)
+	}
+}
+
+func TestConsumer_with_createtime_updatetime(t *testing.T) {
```
Comment (moonming) [inline api/test/e2e/consumer_test.go:155]: Why not use the testing framework?  Is this test related to the title?

**Review by moonming — CHANGES_REQUESTED** [review]

File: api/test/e2e/consumer_test.go (hunk 6 above)
Comment (idbeta) [reply api/test/e2e/consumer_test.go:155]: This case is used to test when the consumer is updated, create_time is not updated, but update_time is updated. The current framework does not support obtaining update_time and create_time from the response, so I use `standard library` and `assert` to achieve it.

**Review by membphis — APPROVED** [review]

File: api/test/e2e/consumer_test.go (hunk 4 above)
Comment (idbeta) [reply api/test/e2e/consumer_test.go]: fixed

File: api/test/e2e/consumer_test.go (hunk 5 above)
Comment (idbeta) [reply api/test/e2e/consumer_test.go]: fixed

File: api/test/e2e/consumer_test.go (hunk 6 above)
Comment (idbeta) [reply api/test/e2e/consumer_test.go:264]: fixed

File: api/test/e2e/consumer_test.go [hunk 7]
```diff
@@ -0,0 +1,253 @@
...
+    }`
+	request, _ := http.NewRequest("PUT", basepath, strings.NewReader(data))
+	request.Header.Add("Authorization", token)
+	resp, _ := http.DefaultClient.Do(request)
+	defer resp.Body.Close()
+	respBody, _ := ioutil.ReadAll(resp.Body)
+	code := gjson.Get(string(respBody), "code")
+	assert.Equal(t, code.String(), "0")
+
+	time.Sleep(1 * time.Second)
```
[the remark this answers is not in the document]
Comment (idbeta) [reply api/test/e2e/consumer_test.go]: It takes more than one second to wait, otherwise, the `update_time` timestamp will not change. because the smallest unit of timestamp here is second.

File: api/test/e2e/consumer_test.go [hunk 8]
```diff
@@ -0,0 +1,250 @@
...
+	resp, _ = http.DefaultClient.Do(request)
+	respBody, _ = ioutil.ReadAll(resp.Body)
+	code = gjson.Get(string(respBody), "code")
+	assert.Equal(t, code.String(), "0")
+
+	time.Sleep(1 * time.Second)
+
+	request, _ = http.NewRequest("GET", basepath+"/jack", nil)
+	request.Header.Add("Authorization", token)
+	resp, _ = http.DefaultClient.Do(request)
```
[the remark this answers is not in the document]
Comment (idbeta) [reply api/test/e2e/consumer_test.go:321]: fixed

File: api/test/e2e/consumer_test.go (hunk 1 above)
Comment (idbeta) [reply api/test/e2e/consumer_test.go]: fixed

File: api/test/e2e/consumer_test.go (hunk 2 above)
Comment (idbeta) [reply api/test/e2e/consumer_test.go]: fixed.

File: api/test/e2e/consumer_test.go (hunk 6 above)
Comment (idbeta) [reply api/test/e2e/consumer_test.go:264]: ping @moonming

File: api/test/e2e/consumer_test.go [hunk 9]
```diff
@@ -0,0 +1,250 @@
...
+	createtime2 := gjson.Get(string(respBody), "data.create_time")
+	updatetime2 := gjson.Get(string(respBody), "data.update_time")
+
+	assert.Equal(t, createtime.String(), createtime2.String())
+	assert.NotEqual(t, updatetime.String(), updatetime2.String())
+
+	//deletea consumer
+	request, _ = http.NewRequest("DELETE", basepath+"/jack", nil)
+	request.Header.Add("Authorization", token)
+	_, err := http.DefaultClient.Do(request)
```
[the remark this answers is not in the document]
Comment (idbeta) [reply api/test/e2e/consumer_test.go]: fixed

File: api/test/e2e/consumer_test.go (hunk 6 above)
Comment (idbeta) [reply api/test/e2e/consumer_test.go:264]: ping @moonming Can you review it when you have time?

File: api/test/e2e/consumer_test.go [hunk 10]
```diff
@@ -0,0 +1,253 @@
...
+	}
+}
+
+func TestConsumer__with_createtime_updatetime(t *testing.T) {
+	//create consumer
+	basepath := "http://127.0.0.1:8080/apisix/admin/consumers"
+	data := `{
+		"username":"jack",
+		"desc": "new consumer"
+    }`
```
[the remark this answers is not in the document]
Comment (idbeta) [reply api/test/e2e/consumer_test.go]: fixed

File: api/test/e2e/consumer_test.go [hunk 11]
```diff
@@ -131,3 +136,219 @@ func TestConsumer_with_key_auth(t *testing.T) {
...
+			ExpectStatus: http.StatusOK,
+			Sleep:        sleepTime,
+		},
+	}
+
+	for _, tc := range tests {
+		testCaseCheck(tc)
+	}
+
+	//get the consumer
```
Comment (membphis) [inline api/test/e2e/consumer_test.go:318]: we should always add one space after `//`

Comment (idbeta) [reply api/test/e2e/consumer_test.go]: fixed.

**Comment by membphis** [comment]
@idbeta CI failed: https://github.com/apache/apisix-dashboard/pull/880/checks?check_run_id=1483626371#step:5:85

it should be a bug.

```
--- FAIL: TestConsumer_with_createtime_updatetime (1.10s)
Error:     printer.go:54: PUT http://127.0.0.1:9000/apisix/admin/consumers
panic: runtime error: invalid memory address or nil pointer dereference [recovered]
	panic: runtime error: invalid memory address or nil pointer dereference
[signal SIGSEGV: segmentation violation code=0x1 addr=0x40 pc=0x8c9e50]

goroutine 11 [running]:
testing.tRunner.func1(0xc000175300)
	/opt/hostedtoolcache/go/1.13.15/x64/src/testing/testing.go:874 +0x3a3
panic(0x956d60, 0xe4e4a0)
	/opt/hostedtoolcache/go/1.13.15/x64/src/runtime/panic.go:679 +0x1b2
e2e.TestConsumer_with_createtime_updatetime(0xc000175300)
	/home/runner/work/apisix-dashboard/apisix-dashboard/api/test/e2e/consumer_test.go:291 +0x350
testing.tRunner(0xc000175300, 0xa1a818)
	/opt/hostedtoolcache/go/1.13.15/x64/src/testing/testing.go:909 +0xc9
created by testing.(*T).Run
	/opt/hostedtoolcache/go/1.13.15/x64/src/testing/testing.go:960 +0x350
exit status 2
FAIL	e2e	1.893s
Error: Process completed with exit code 1.

**Comment by membphis** [comment]
@moonming @nic-chen please take a look at this PR

**Comment by idbeta** [comment]
> @idbeta CI failed: https://github.com/apache/apisix-dashboard/pull/880/checks?check_run_id=1483626371#step:5:85
Because the listening port of `manager-api` has changed, I have added new judgments.

**Comment by membphis** [comment]
@moonming please review this PR if you have time

@idbeta please change the title, your test case is not the same as this title.

**Review by moonming — APPROVED** [review]

**Comment by nic-chen** [comment]
LGTM
